### ISSUE TITLE
[Regression] Several PSN(HDD) games are unable to boot from an external directory

### ISSUE THREAD

**Issue by Rex000** [RPCS3/rpcs3#15027]
### Quick summary

Several PSN(HDD) games are unable to boot from an external directory.

### Details

I always keep my games, hdd or disc, outside of the rpcs3 folder in another directory. I use "Add Games" feature to add them to the games list, and they were working flawlessly. Now in the latest build I found out that many of those games failed to boot.

Upon inspection of several older builds, I narrowed it down to build # 0.0.29-15856 [#14956](https://github.com/RPCS3/rpcs3/pull/14956)

### Build with regression

v0.0.29-15856

### Attach two log files

Working build:
[RPCS3v0.0.29-15855.log.zip](https://github.com/RPCS3/rpcs3/files/13868484/RPCS3v0.0.29-15855.log.zip)

Regression build:
[RPCS3v0.0.29-15856.log.zip](https://github.com/RPCS3/rpcs3/files/13868486/RPCS3v0.0.29-15856.log.zip)

### Other details

**Background:**
This very useful and convenient feature of adding games from an external directory has been available for a while in rpcs3. It had a slight issue #13383 that I reported in February and it was fixed. Have been using it without any issues until now.

**Comment by Megamouse** [comment]
Not a bug. The "feature" of having hdd games in random external dirs was only ever a hack to begin with and was never truly intended. We have a VFS for that exact reason which is drastically undermined by this behaviour. But I guess it's too late to complain.

The key problem with these external dirs is that you have to keep the correct foldername in which the game was installed in, because the games access these folders (sometimes hardcoded) by name and not by any information that we can generalize.
We did exactly that in the past by using the title id, which turned out to be wrong.
So when you renamed the directory from e.g. "NPEB12345" or "NPEB12345MYGAME" to "Game name NPEB12345" you unknowingly made it incompatible with the newer (and correct) code.

You can fix this by either renaming those folders to the original name or (for most games) by simply putting the folder content into a subdirectory named the same as the title id. You then may have to add the games again.

**Comment by Rex000** [comment]
Ok here are some games I tried with your naming schemes.

**Hyperballoid HD:**
NPUB30189Hyperballoid HD -- didn't work
Hyperballoid HD -- didn't work
NPUB30189 -- worked
Hyperballoid HD NPUB30189 -- didn't work

**Super Stardust HD:**
NPUA80068Super Stardust HD -- worked
Super Stardust HD -- worked
NPUA80068  -- worked
Super Stardust HD NPUA80068  -- worked

**After Burner Climax:**
NPUB30151After Burner Climax -- worked
After Burner Climax -- worked
NPUB30151 -- worked
After Burner Climax NPUB30151 -- worked

**TEKKEN TAG TOURNAMENT 2:**
NPEB01140TEKKEN TAG TOURNAMENT 2 -- didn't work
TEKKEN TAG TOURNAMENT 2 -- worked

**Comment by Rex000** [comment]
Ok, here are some games I tried with your given naming schemes.

**Hyperballoid HD:**
NPUB30189Hyperballoid HD -- didn't work
Hyperballoid HD -- didn't work
NPUB30189 -- worked
Hyperballoid HD NPUB30189 -- didn't work

**Super Stardust HD:**
NPUA80068Super Stardust HD -- worked
Super Stardust HD -- worked
NPUA80068  -- worked
Super Stardust HD NPUA80068  -- worked

**After Burner Climax:**
NPUB30151After Burner Climax -- worked
After Burner Climax -- worked
NPUB30151 -- worked
After Burner Climax NPUB30151 -- worked

**TEKKEN TAG TOURNAMENT 2:**
NPEB01140TEKKEN TAG TOURNAMENT 2 -- didn't work
TEKKEN TAG TOURNAMENT 2 -- worked
NPEB01140 -- worked
TEKKEN TAG TOURNAMENT 2 NPEB01140 -- didn't work

**Bionic Commando Rearmed 2:**
NPUB30131Bionic Commando Rearmed 2 -- didn't work
Bionic Commando Rearmed 2 -- didn't work
NPUB30131 -- worked
Bionic Commando Rearmed 2 NPUB30131 -- didn't work

What is this inconsistency?

>You can fix this by either renaming those folders to the original name or (for most games) by simply putting the folder content into a subdirectory named the same as the title id. You then may have to add the games again.

I tried this as well. But it's not allowing me to add one directory with all the game folders inside it, because it seemingly doesn't detect the subdirectories named as the title id's inside each game folder. I have to add every individual game separately which obviously isn't too convenient.

So basically, it's not detecting any games when I try to add the directory "F:\PS3Games" with this folder structure:
F:\PS3Games\Bionic Commando Rearmed 2\NPUB30131
F:\PS3Games\Hyperballoid HD\NPUB30189
F:\PS3Games\Super Stardust HD\NPUA80068
F:\PS3Games\TEKKEN TAG TOURNAMENT 2\NPEB01140
F:\PS3Games\After Burner Climax\NPUB30151

**Comment by Megamouse** [comment]
The randomness is what i explained earlier: some games go by cellGame dir and some games hardcode to the known dir
And since we only know this dir by installing the game, we can't generalize by using title id.
So some games may work with the wrong name and others don't. Therefore you shouldn't rename them in the first place.

The adding of one directory with double nested directories not working is expected, since we only scan the first directory level for game content. You can just drag and drop all the directories below F:\PS3Games into rpcs3 and they should be added anyway.

**Comment by Rex000** [comment]
> The adding of one directory with double nested directories not working is expected, since we only scan the first directory level for game content. You can just drag and drop all the directories below F:\PS3Games into rpcs3 and they should be added anyway.

Yes, that seems to work.

Though, would it be possible to have a "Add Games (include subdirectories)" option in rpcs3? Would be a very convenient thing to have. Also, this information about the specific directory structure should be added to the wiki as well imo, since there are a lot of people who keep their games separate from the emulator directory and won't know what to do to make them work.

**Comment by Rex000** [comment]
> The adding of one directory with double nested directories not working is expected, since we only scan the first directory level for game content. You can just drag and drop all the directories below F:\PS3Games into rpcs3 and they should be added anyway.

Yes, that seems to work.

This information about specific directory structure and naming should be added to the wiki as well imo, since there are a lot of people who keep their games separate from the emulator directory and won't know what to do to make them work.

Though, would it be possible to have a "Add Games (include subdirectories)" option in rpcs3? Would be a very convenient thing to have especially from the average users' standpoint since they won't have a clue why their games are not working despite naming the directories correctly.

**Comment by Rex000** [comment]
> The adding of one directory with double nested directories not working is expected, since we only scan the first directory level for game content. You can just drag and drop all the directories below F:\PS3Games into rpcs3 and they should be added anyway.

Yes, that seems to work.

This information about specific directory structure and naming should be added to the wiki as well imo, since there are a lot of people who keep their games separate from the emulator directory and won't know what to do to make them work.

Also, would it be possible to have a "Add Games (include subdirectories)" option in rpcs3? Would be a very convenient thing to have especially from the average users' standpoint since they won't have a clue why their games are not working despite naming the directories correctly.

**Comment by Rex000** [comment]
One more thing. Could there be a way to add a foolproof mechanism so that the PSN(HDD) games don't get detected unless they have the proper folder naming? I was able to test those five games above with different naming schemes because they kept appearing in the game list when they shouldn't have.

**Comment by Megamouse** [comment]
No. sadly not.
